colibre/scripts/stellar_abundances_FeH_CFe_APOGEE.py

```diff
@@ -0,0 +1,157 @@
+"""
+Plots the stellar abundances ([Fe/H] vs [C/Fe]) for a given snapshot
+"""
+import matplotlib.pyplot as plt
+import matplotlib.patheffects as pe
+import matplotlib.colors as mc
+import numpy as np
+import unyt
+import glob
+from swiftsimio import load
+from swiftpipeline.argumentparser import ScriptArgumentParser
+from velociraptor.observations import load_observations
+from unyt import unyt_array
+
+
+def read_data(data):
+    """
+    Grabs the data
+    """
+
+    mH_in_cgs = unyt.mh
+    mFe_in_cgs = 55.845 * unyt.mp
+    mC_in_cgs = 12.0107 * unyt.mp
+
+    # Asplund et al. (2009)
+    Fe_H_Sun_Asplund = 7.5
+    C_H_Sun_Asplund = 8.43
+
+    C_Fe_Sun = C_H_Sun_Asplund - Fe_H_Sun_Asplund - np.log10(mFe_in_cgs / mC_in_cgs)
+    Fe_H_Sun = Fe_H_Sun_Asplund - 12.0 - np.log10(mH_in_cgs / mFe_in_cgs)
+
+    carbon = data.stars.element_mass_fractions.carbon
+    iron = data.stars.element_mass_fractions.iron
+    hydrogen = data.stars.element_mass_fractions.hydrogen
+
+    Fe_H = np.log10(iron / hydrogen) - Fe_H_Sun
+    C_Fe = np.log10(carbon / iron) - C_Fe_Sun
+
+    Fe_H[iron == 0] = -4  # set lower limit
+    Fe_H[Fe_H < -4] = -4  # set lower limit
+
+    C_Fe[iron == 0] = -2  # set lower limit
+    C_Fe[carbon == 0] = -2  # set lower limit
+    C_Fe[C_Fe < -2] = -2  # set lower limit
+
+    return Fe_H, C_Fe
+
+
+arguments = ScriptArgumentParser(
+    description="Creates an [Fe/H] - [C/Fe] plot for stars."
+)
+
+snapshot_filenames = [
+    f"{directory}/{snapshot}"
+    for directory, snapshot in zip(arguments.directory_list, arguments.snapshot_list)
+]
+
+names = arguments.name_list
+output_path = arguments.output_directory
+
+plt.style.use(arguments.stylesheet_location)
+
+simulation_lines = []
+simulation_labels = []
+
+fig, ax = plt.subplots()
+ax.grid(True)
+
+for snapshot_filename, name in zip(snapshot_filenames, names):
+
+    data = load(snapshot_filename)
+    redshift = data.metadata.z
+
+    Fe_H, C_Fe = read_data(data)
+
+    # Bins along the X axis (Fe_H) to plot the median line
+    bins = np.arange(-4.1, 1, 0.2)
+    ind = np.digitize(Fe_H, bins)
+
+    xm, ym, ym1, ym2 = [], [], [], []
+    Min_N_points_per_bin = 11
+
+    for i in range(1, len(bins)):
```

Comment:
I have some comments about this binning strategy.

I'm not quite sure it is correct to use the median for both the x and y variable in a bin. When you bin quantities like this, then the x variable is treated as an independent variable for which you impose a value (the bin centre) and an associated uncertainty (the bin width). The y variable is then a dependent variable that has a median and uncertainty given those values of the independent variable. By using the median for x as well, you arbitrarily shift the x values within their uncertainty interval (the bin), without really adding any useful information.

Manually binning values like this is also really slow compared to using `scipy.stats.binned_statistic`. You can get the same curve (without medians in x) using
```
xm = 0.5 * (bins[1:] + bins[:-1])
ym,_,_ = stats.binned_statistic(Fe_H, C_Fe, statistic="median", bins=bins)
ym1,_,_ = stats.binned_statistic(Fe_H, C_Fe, statistic=lambda x: np.percentile(x, 16.), bins=bins)
ym2,_,_ = stats.binned_statistic(Fe_H, C_Fe, statistic=lambda x: np.percentile(x, 84.), bins=bins)
counts,_,_ = stats.binned_statistic(Fe_H, C_Fe, statistic="count", bins=bins)
mask = counts >= Min_N_points_per_bin
xm = xm[mask]
ym = ym[mask]
ym1 = ym1[mask]
ym2 = ym2[mask]
```

Reply:
I followed the median binning, since this was how it was originally implemented in stellar_abundances_FeH_MgFe.py in the main branch of the pipeline-config/colibre. But I can update it to use scipy.